src/App.js

```diff
@@ -30,42 +33,98 @@ const App = () => {
   // This starts state off as a 2D array of JS objects with
   // empty value and unique ids.
   const [squares, setSquares] = useState(generateSquares());
+  const [currentPlayer, setCurrentPlayer] = useState(PLAYER_1);
+  const [winner, setWinner] = useState('');
+  const [turns, setTurns] = useState(0)
 
   // Wave 2
   // You will need to create a method to change the square 
   //   When it is clicked on.
   //   Then pass it into the squares as a callback
+  const onClickCallback = (id) => {
+    let row = 0
+    while (id > 2) {
+      id -= 3
+      row++
+    }
 
+    if (!squares[row][id]['value'] && !winner) {
+      const newSquares = [...squares]
+      newSquares[row][id]['value'] = currentPlayer
+      setSquares(newSquares)
 
-  const checkForWinner = () => {
+      const updatedPlayer = currentPlayer === PLAYER_1 ? PLAYER_2 : PLAYER_1
+      setCurrentPlayer(updatedPlayer)
+      const gameOver = checkForWinner(row, id)
+      if (!gameOver) {
+        if (turns === 8) {
+          setWinner('Tie')
+        } else {
+          setTurns(turns + 1)
+        }
+      }
+    }
+  }
+
+  const checkForWinner = (row, col) => {
     // Complete in Wave 3
-    // You will need to:
-    // 1. Go accross each row to see if 
-    //    3 squares in the same row match
-    //    i.e. same value
-    // 2. Go down each column to see if
-    //    3 squares in each column match
-    // 3. Go across each diagonal to see if 
-    //    all three squares have the same value.
 
+    // check row
+    if (squares[row][0]['value'] === currentPlayer &&
+      squares[row][1]['value'] === currentPlayer &&
+      squares[row][2]['value'] === currentPlayer) {
+      setWinner(currentPlayer)
+      return true
+    }
+
+    // check col
+    if (squares[0][col]['value'] === currentPlayer &&
+      squares[1][col]['value'] === currentPlayer &&
+      squares[2][col]['value'] === currentPlayer) {
+      setWinner(currentPlayer)
+      return true
+    }
+
+    // check diag
+    if ((squares[0][0]['value'] === currentPlayer &&
+      squares[1][1]['value'] === currentPlayer &&
+      squares[2][2]['value'] === currentPlayer)
+      || (squares[0][2]['value'] === currentPlayer &&
+        squares[1][1]['value'] === currentPlayer &&
+        squares[2][0]['value'] === currentPlayer)
+    ) {
+      setWinner(currentPlayer)
+      return true
+    }
+    return false
   }
 
+
   const resetGame = () => {
-    // Complete in Wave 4
+    setSquares(generateSquares())
+    setCurrentPlayer(PLAYER_1)
+    setWinner('')
+    setTurns(0)
   }
 
+  const gameResults = winner === 'Tie' ? 'Tie Game' :
+    winner === PLAYER_1 ? `Winner is ${PLAYER_1}` : `Winner is ${PLAYER_2}`
```

Comment:
Try to avoid nested ternaries. Splitting the lines and the indentation helps, but if you really want to avoid having if conditions _here_, I'd suggest a helper function that you could call here.